File: src/particles/Computer.cpp
#include "particles/Computer.hpp"
#include "storage/Storage.hpp"
#include "storage/PropertyHandle.hpp"
#include "Property.hpp"
#include "python.hpp"

using namespace espresso;
using namespace espresso::storage;
using namespace espresso::particles;
using namespace espresso::python;

namespace espresso {
  namespace particles {
    class PythonComputer
      : public Computer,
	public wrapper< Computer > {
      
    public: 
      void prepare(Storage::SelfPtr storage) {
	// get the particleId property
	particleId = storage->getIdPropertyHandle();
    
	// call Python prepare
	if (override prepare = get_override("prepare"))
	  prepare(storage);
      }

      /** The operator() calling the Python objects function each */
      bool apply(const ParticleHandle pref) {
	return get_override("apply")(particleId[pref]);
      }
  
      void finalize() {
	// call Python finalize
	if (override finalize = get_override("finalize"))
	  finalize();
	else Computer::finalize();
      }

    private:
      PropertyHandle< ParticleId > particleId;
    };
  }
}

void Computer::registerPython() {
  espresso::python::class_
    < espresso::particles::PythonComputer, boost::noncopyable >
    ("particles_PythonComputer")
    .def("prepare", pure_virtual(&Computer::prepare))
    .def("finalize", &Computer::finalize)
    .def("apply", pure_virtual(&Computer::apply))
    ;

}

File: src/particles/Computer.py
from espresso import pmi
from espresso.esutil import cxxinit
from _espresso import particles_PythonComputer

# Python base class for Computers
class PythonComputerLocal(particles_PythonComputer):
    """
    a per-particle-computer using Python code. You mainly need to
    provide a function "apply", so that a
    PythonComputerLocal-object can be applied to a particle
    (identity). "apply" is mandatory to override. If any of two
    other functions is provided, namely "prepare" and/or "finalize",
    they are called right before calling "apply" and right after,
    respectively. The "prepare" function receives as input parameter a
    DecomposerLocal-object, the one that the subsequent
    "apply"-calls will operate on. Several calls to "apply"
    might occur before the call to "finalize", but the
    Decomposer-object will not change inbetween.

    For example, the Decomposer.foreach-method will first call
    "prepare" on each node, handing over its local instance. Then on
    each node, it will call "apply" for each particle, and after
    that, call "finalize" once on each node. In the case of
    Decomposer.foreach, the return value of "finalize" is passed back.
    """
    def __init__(self):
        cxxinit(self, particles_PythonComputer)
        
    def prepare(self, storage):
        pass

    def apply(self, id): 
        """
        override this function to implement a per-particle computer. This routine will be
        called for each particle, with id being the identity of the particle.
        """
        pass

    def finalize(self): pass

# ABC for worker Computers
class ComputerLocal(object):
    pass

if pmi.IS_CONTROLLER:
    # ABC for controller Computers
    class Computer(object):
        __metaclass__ = pmi.Proxy
